File: atcoder/abc/045/c.cpp
///
// File:  c.cpp
// Author: ymiyamoto
//
// Created on Sun Oct  1 21:09:59 2017
//

#include <cstdint>
#include <iostream>
#include <map>
#include <string>

using namespace std;
#define SIZE 10

static bool sep[SIZE];
static string str;

static uint64_t calc_count()
{
  uint64_t count = 0;
  uint64_t val = 0;
  for (uint32_t i = 0; i < str.size(); i++) {
    val *= 10;
    val += str[i] - '0';
    if (sep[i]) {
      count += val;
      val = 0;
    }
  }
  count += val;
  return count;
}

static uint64_t perm(uint32_t i, uint32_t N)
{
  uint64_t count = 0;

  if (i < N) {
    sep[i] = true;
    count += perm(i + 1, N);
    sep[i] = false;
    count += perm(i + 1, N);
  } else {
    count += calc_count();
  }

  return count;
}

int32_t main()
{
  cin >> str;

  cout << perm(0, str.size() - 1) << endl;

  return 0;
}

File: atcoder/abc/045/c/c.go
///
// File:  c.go
// Author: ymiyamoto
//
// Created on Tue Mar  6 22:15:13 2018
//
package main

import (
	"fmt"
)

var N, M int
var graph [][]bool

func permutations(n int) [][]int {
	if n == 1 {
		return [][]int{{1}}
	}

	ret := [][]int{}
	perm := permutations(n - 1)
	for i := 0; i < len(perm); i++ {
		for j := 0; j <= len(perm[i]); j++ {
			p := make([]int, len(perm[i]))
			copy(p, perm[i])
			ret = append(ret, append(p[:j], append([]int{n}, p[j:]...)...))
		}
	}
	return ret
}

func main() {
	fmt.Scan(&N, &M)

	graph = make([][]bool, N+1)
	for i := 0; i < N+1; i++ {
		graph[i] = make([]bool, N+1)
	}

	for i := 0; i < M; i++ {
		var a, b int
		fmt.Scan(&a, &b)
		graph[a][b] = true
		graph[b][a] = true
	}

	count := 0
	for _, val := range permutations(N) {
		if val[0] == 1 {
			invalid := false
			for i := range val {
				if i+1 < len(val) {
					if !graph[val[i]][val[i+1]] {
						invalid = true
					}
				}
			}
			if !invalid {
				count++
			}
		}
	}
	fmt.Println(count)
}
